Update project configuration and pods to build for Apple Silicon

mokagio opened the pull request:
### Fix
Jorge noticed that the app wasn't built for Apple Silicon, a massive fail on my behalf :flushed:

This PR updates a bunch of things, all unfortunately required to make the build happen.

- CocoaPods to version 1.10.1 because that has proper support for `.xcframework`s, required by...
- Sodium to version 0.9.1, which has support for Apple Silicon and is included in
- Tracks, which also includes
- Sentry (via @jkmassel's PR https://github.com/Automattic/Automattic-Tracks-iOS/pull/166/) because the current version has a macro to exclude iOS specific code when not building for arm64, which no longer works because we're now building for arm64 on macOS, too

Finally, the `CrashLogging` implementation had to be tweaked, too, in order to compile with the Sentry-related changes in Tracks.

### Merge Thoughts

~~I based this PR off `release/2.6` hoping that it can make it in there, but **it would be appropriate to wait till an official version of Tracks has been released before merging this PR.**~~

~~That might mean delaying shipping version 2.6 to the users for a few days, or longer in case there are issues with the Sentry migration.~~

~~If the work on Tracks takes longer, we can shift the Apple Silicon support to version 2.7 and have a more thorough testing of it.~~

This is now based off `release/2.7`.

### Test

Checkout this code, run `bundle install && bundle exec pod install`, build then archive the Simplenote-AppStore scheme. Run the app and make sure it works. ~~Do the same for the Simplenote scheme.~~ _See #803 for that_.

<img width="1112" alt="Screen Shot 2021-01-20 at 3 07 57 pm" src="https://user-images.githubusercontent.com/1218433/105128029-417ced80-5b36-11eb-9bb4-bac66c9acf0f.png">


### Review
I'd like a review from a MIE and from someone with an M1 Mac to verify it builds fine there.

### Release
`RELEASE-NOTES.txt` was updated in 36f3b2b with:

> Add Apple Silicon build support

mokagio committed:
Remove VALID_ARCHS build setting to enable Apple Silicon destination

This option does not appear in a new project generated with Xcode 12.0.
Removing it rather than updating it to include arm64 seemed like a
cleaner approach.

See discussion here:
https://developer.apple.com/forums/thread/651181

mokagio committed:
Use Ruby version 2.6.4 - The default across our current tooling stack

mokagio committed:
Update CocoaPods to the latest version - 1.10.1

To do so, we also had to update the release toolkit Fastlane plugin
because of shared dependencies resolution.

mokagio committed:
Update project file to Xcode 12.0 compatible format

This will enable a neater integration with CocoaPods, using a file list
for the frameworks rather than listing them directly in the project.

mokagio committed:
Switch to WIP branch on Tracks with Apple Silicon support

Notice that to keep this commit atomic, a few changes to the
implementation had to be included to ensure the build succeeded.

mokagio committed:
Update release notes: Add Apple Silicon build support

mokagio commented on Gemfile:
```diff
@@ -1,7 +1,6 @@
   gem 'rake'
-  gem 'cocoapods', '~> 1.8'
-  gem 'cocoapods-repo-update', '~> 0.0.3'
+  gem 'cocoapods', '>= 1.10'
```
CocoaPods 1.10.1 is required to properly build the `xcframework`s that Sodium (a Tracks dependency) uses to support Apple Silicon.

The repo update plugin blocked the dependency resolution and it should anyways no longer necessary because CocoaPods uses the CDN by default.

mokagio commented on Podfile:
```diff
@@ -6,16 +6,13 @@ use_frameworks!
   # Automattic Shared
   #
-  pod 'Automattic-Tracks-iOS', '~> 0.6'
+  pod 'Automattic-Tracks-iOS', git: 'git@github.com:Automattic/Automattic-Tracks-iOS', branch: 'update/sentry-to-v6'
```
**It would be appropriate to wait till an official version of Tracks has been released before merging this PR.**

That might mean delaying shipping version 2.6 to the users for a few days, or longer in case there are issues with the Sentry migration. 

Alternatively, we can shift the Apple Silicon support to version 2.7 and have a more thorough testing of it.

jleandroperez replied:
Good on my end! We can switch when it's available!

jkmassel replied:
I've switched this over

mokagio commented on Simplenote.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 52;
+	objectVersion = 54;
```
This and the `Xcode 12.0` change below are so that CocoaPods can keep using a file to provide the frameworks list as part of its integration. Otherwise, it will list them directly in the project, making for noisier diffs.

mokagio commented on Simplenote.xcodeproj/project.pbxproj:
```diff
@@ -1762,7 +1762,7 @@
 			runOnlyForDeploymentPostprocessing = 0;
 			shellPath = /bin/sh;
-			shellScript = "plist=\"${TARGET_BUILD_DIR}/${INFOPLIST_PATH}\"\nyear=$(date \"+%Y\")\ncopyright=\"Copyright © $year Automattic. All rights reserved.\"\n\n# Set the version numbers in the buildPlist\n/usr/libexec/PlistBuddy -c \"Set :NSHumanReadableCopyright $copyright\" $plist\n";
+			shellScript = "plist=\"${TARGET_BUILD_DIR}/${INFOPLIST_PATH}\"\nyear=$(date \"+%Y\")\ncopyright=\"Copyright © $year Automattic. All rights reserved.\"\n\n# Set the version numbers in the buildPlist\n#\n# Temporarily disabled! It's failing the build after updating to support Apple Silicon.\n#\n# Unfortunately, a number of changes were required to make the build pass before getting to this point ...
```
You'll have to scroll horizontally a bit to see the change. Here's a formatted version:

```diff
  plist="${TARGET_BUILD_DIR}/${INFOPLIST_PATH}"
  year=$(date "+%Y")
  copyright="Copyright © $year Automattic. All rights reserved."

  # Set the version numbers in the buildPlist
- # /usr/libexec/PlistBuddy -c "Set :NSHumanReadableCopyright $copyright" $plist
+ #
+ # Temporarily disabled! It's failing the build after updating to support Apple Silicon.
+ #
+ # Unfortunately, a number of changes were required to make the build pass before getting to this point,
+ # so there is no clear responsible for this failure yet.
+ # /usr/libexec/PlistBuddy -c "Set :NSHumanReadableCopyright $copyright" $plist
```

I'd be curious to see if this happens for you, too.

jkmassel replied:
This doesn't happen to me locally [or in CI](https://app.circleci.com/pipelines/github/Automattic/simplenote-macos/1830/workflows/9ab7856b-4399-4f2e-8b2a-2605425b52f4/jobs/1784), so AFAICT it's safe to put back, but LMK if not?

mokagio commented on Simplenote.xcodeproj/project.pbxproj:
```diff
@@ -2409,7 +2409,6 @@
 				SDKROOT = macosx;
 				SWIFT_OPTIMIZATION_LEVEL = "-Onone";
 				SWIFT_VERSION = 5.0;
-				VALID_ARCHS = x86_64;
```
This appears to be a legacy build setting: if you create a new macOS with Xcode 12.0, it won't be there.

I opted for removing it instead of updating it to include `arm64`.

jleandroperez replied:
Apparently it defaults to `Standard Archs` so it's all good! thank you!!

mokagio commented on Simplenote/TagListState.swift:
```diff
@@ -108,7 +108,7 @@ extension TagListState {
     @objc
     func tag(atIndex index: Int) -> Tag? {
-        guard case let .tag(tag) = rowAtIndex(index) else {
+        guard let row = rowAtIndex(index), case let .tag(tag) = row else {
```
The compiler required `rowAtIndex(_:)` `Optional` return value to be unwrapped, although I don't understand why this failure appeared now, since it's not related with Tracks 🤔

jleandroperez replied:
Absolutely a mystery to me as well, this change is unrelated to anything else. I suspect it's perhaps a glitch in the compiler (since one of the changes is the supported archs right?).

No big deal, it's a valid change!! thank you Gio!!

AliSoftware replied:
I think that's a change in a recent Swift compiler update?

Some other possibilities to fix it would have been to use `?` to match with the proper optional lhs type:
```swift
        // '?' shorthand syntax for matching `Optional.some`
        guard case let .tag(tag)? = rowAtIndex(index) else {
        // with the `let` as close as possible to the variable you want to bind (which I personally prefer, but 🤷 )
        guard case .tag(let tag)? = rowAtIndex(index) else {
```

Which are both valid shorthands for the following more expanded forms of `Optional.some` (since that's an non-nil optional we're trying to match with here) if that helps see it better:
```swift
        // Fully expanded form with fully qualified names
        guard case Optional.some(TagListRow.tag(let tag)) = rowAtIndex(index)
        // Explicit Optional case
        guard case .some(.tag(let tag)) = rowAtIndex(index)
```

I think in previous versions of Swift the compiler would allow automatically lifting the operands to optionals if needed but that got removed to avoid ambiguity found in some cases.

jkmassel commented:
I'm happy to review this since most of it is the Tracks update, but if it's ok with y'all I'll wait until the Tracks PR lands – at that point I'd love to test this a bunch to ensure we're good to go!

jleandroperez commented:
@mokagio code changes look good!! But I'm having issues archiving the `AppStore` target on my end. Will poke you later today!!

```
/Users/lantean/Developer/Automattic/simplenote-macos/Simplenote/Note.h:9:9: error: module 'Simperium_OSX' not found
@import Simperium_OSX;
        ^
```

mokagio commented:
@jleandroperez and I worked on this together earlier.

Archiving worked on my machine but not on his. We noticed he's on Big Sur and I'm on still on Catalina so we thought the issue was due to that. I tried on my other Mac with Big Sur and... bang! I got the same issue: building works but archiving doesn't 🤔 

Only problem was... I wasn't sure how to fix it.

After some poking around I noticed that I was archiving for "My Mac". Once I switched to "Any Mac (Apple Silicon, Intel)" the archive worked 🎉 🤦 

I went back to my main Catalina machine and – _you guess it_ – the destination was Any Mac. If I switch it to My Mac, the archive fails.

<img width="1512" alt="Screen Shot 2021-01-21 at 2 15 34 pm" src="https://user-images.githubusercontent.com/1218433/105275748-18229700-5bf4-11eb-8691-7a590dd2f240.png">


Here's an excerpt from the [Xcode documentation](https://developer.apple.com/documentation/xcode/distributing_your_app_for_beta_testing_and_releases#3405659):

> In the main window of your Xcode project, choose a target and a build-only device or real device from the Scheme toolbar menu. If the target is a simulator, you can’t create an archive. Then, choose Product > Archive to build the target, and create the archive that will appear in the Archives organizer.

It's super easy thing to overlook: on iOS, Xcode disables the Archive menu item when the destination is a Simulator, but that's not the case for Mac (I'm guessing because "My Mac" counts as a real device?)

I'm still unsure why the error is that the Simperium module can't be found with destination "My Mac" 🤔 But, unless archiving for "Any Mac" still fails for you @jleandroperez, I think we can consider that issue closed.

mokagio commented:
> I'm happy to review this since most of it is the Tracks update, but if it's ok with y'all I'll wait until the Tracks PR lands – at that point I'd love to test this a bunch to ensure we're good to go!

Sounds good to me 👍

jleandroperez reviewed (approved):
Thank you @mokagio, that was it, `Any Mac`!!!

Good to go sir!!

:shipit: :shipit: :shipit:

mokagio commented:
Moved this to 2.7 after discussing it with @jleandroperez. Since the Tracks work is still in flux, it would be unsafe to ship it to the users. M1 owners can still enjoy Simplenote because of Rosetta.

jkmassel committed:
Build for Sparkle in debug non-App Store builds

This will prevent it from randomly breaking when we try to do a production build

jkmassel committed:
Use Tracks 0.8.0

jkmassel commented on .ruby-version:
```diff
@@ -1 +1 @@
-2.5.5
+2.6.4
```
Do we need to use a newer Ruby version for this repo?

mokagio replied:
No we don't. I updated this to keep it in line with the rest of our projects (b79d8d2587810cf59124da7d65eb6863d124cdf8).

Do you think it's a problem? An out-of-scope change on a PR that should be as little as possible given it targets the release branch?

AliSoftware replied:
I personally would keep that change, as I agree it's nice to use the same ruby version across all our tooling.
I did a similar update in the past on another PR and repo (can't remember which one, but we discussed it in PR comments with Gio at that time).

The benefits of using the same ruby across our repos whenever possible is to avoid having to `rbenv install` and `bundle install` all the gems again (duplicating all the downloads and size on disk for the gems for usually no main benefit), while I don't really see any major drawback.

jkmassel replied:
kk works for me if we're making things _more_ consistent!

jkmassel commented on Gemfile.lock:
```diff
@@ -297,4 +300,4 @@ DEPENDENCIES
 
 BUNDLED WITH
-   2.1.4
+   2.2.5
```
Did we mean to change this?

mokagio replied:
I did it intentionally, yes. The reason being that Bundler 2.2.x fixes an issue that may result in confusing changes to the `.bundle/config` file (more details in this Simplenote iOS PR https://github.com/Automattic/simplenote-ios/pull/1059).

Do you think it's a problem? An out-of-scope change on a PR that should be as little as possible given it targets the release branch?

jkmassel replied:
I'm good with it! Just wanted to be sure it was intentional!

jkmassel commented on Simplenote/CrashLogging.swift:
```diff
@@ -7,19 +7,22 @@ import AutomatticTracks
 @objc(CrashLogging)
 class CrashLoggingShim: NSObject {
+
+    private static var sharedCrashLogging: CrashLogging?
```
Instead of the singleton here, could we instead move this to the `AppDelegate` and explicitly initialize it there?

All of the calls into this shim happen from the AppDelegate anyway, so there they could reference the instance, and we could get rid of all the `static` business going on here?

mokagio replied:
Great suggestion, thank you for the extra context. I'll admit I went with this option simply because it allowed me to keep the change in this file without touching more of the app.

I'll give your suggestion a try, as I'm not happy with singletons either.

jkmassel reviewed (changes requested):
I updated this PR to use Tracks 0.8.0 prior to testing – the changes are all in 6f95235cf776b9ca2f8ea8ddf37a97ee2550e624. While I was there, I also dropped the old Pods workaround.

A few things that caught my eye:
- Some (possibly?) unintended changes to the Ruby setup
- Ideally we wouldn't introduce another singleton object in the CrashLogging shim 😅

mokagio committed:
Update Sparkle feed URL to use App Center format

This was needed anyways but became required as, in moving from version
1.22 manually added to the project to 1.24 linked via CocoaPods, the
original Hockey URL stopped working.

mokagio committed:
Use CrashLoggingShim via instance instead of statically

Thanks @jkmassel for the suggestion.

See discussion at https://github.com/Automattic/simplenote-macos/pull/802/files#r564803911

mokagio committed:
Merge branch 'release/2.7' into release/2.6-apple-silicon

Conflicts:

- `Gemfile.lock`
- `Podfile`
- `Podfile.lock`
- `Simplenote/SimplenoteAppDelegate.m`
- `fastlane/Pluginfile`

mokagio commented on Simplenote/CrashLogging.swift:
```diff
@@ -8,21 +8,32 @@ import AutomatticTracks
 class CrashLoggingShim: NSObject {
 
-    private static var sharedCrashLogging: CrashLogging?
+    private let crashLogging: CrashLogging
```
@jkmassel from your suggestion, I got the idea of fully removing this class and calling its methods directly in the application delegate.

That's not possible because `CrashLogging` is a Swift type that's not visible to the Objective-C `AppDelegate`.

This is an attempt to remove the `static` component while staying within the Objective-C constraints. I created a private instance of the Swift `CrashLogging` that this shim calls via its instance methods visible to the Objective-C code.

AliSoftware replied:
Wondering if we couldn't make the main `CrashLogging` Swift type be `@objc`-compatible in the first place?

mokagio replied:
Thank you for the suggestion. 🙇 

`CrashLogging` comes from Tracks, so updating it would require another round trip. That's not a huge issue, just a possible time sink that might prevent us from landing this in `release/2.7`.

I'd be reluctant to go down that path, though, because I fear a rabbit hole. @jkmassel you have  more experience with that code, so maybe you can  tell me if I'm being overly cautious?

Maybe we should update the app delegate to be Swift, instead 😛 I'm only half joking. Obviously that would be a bigger piece of work, but it might a good time investment to modernize the codebase?

jkmassel replied:
Thanks for looking at this @mokagio. @AliSoftware – we could certainly do that, though Simplenote is the only Objective-C (iOS and macOS) App Delegate (and ideally IMHO wouldn't own the Tracks `CrashLogging` class directly). 

FWIW it wouldn't be that hard to do, just need to conform to `NSObject`.

mokagio committed:
Re-enable and tidy up Build Phase to set copyright year

Thanks @jkmassel; it works for me now, too.

I just updated to Xcode 12.4, but I doubt that's the reason because it
wasn't out yet at the time of the commit that disabled this. Possibly
some local messed up state in my Xcode that even removing the
DerivedData folder didn't solve. 🤷‍♂️

jkmassel commented on Simplenote/SimplenoteAppDelegate.m:
```diff
@@ -163,7 +165,7 @@ - (void)handleGetURLEvent:(NSAppleEventDescriptor *)event withReplyEvent:(NSAppl
 - (void)setupCrashLogging
 {
-    [CrashLogging startWithSimperium: self.simperium];
+    self.crashLogging = [[CrashLogging alloc] initWithSimperium:self.simperium];
```
```suggestion
    self.crashLogging = [[CrashLogging alloc] initWithSimperium:self.simperium];
    [self.crashLogging start];
```

We need to call `start` on the `CrashLogging` object in order to initialize the SDK internally.

mokagio replied:
🤦‍♂️

AliSoftware replied:
If it's a common pattern it could have been nice to add a `static func start(with simperium: Simperium)` function in the lib that would be just a wrapper around `init(simperium:)` + `start()` just for nicer API… bwah, too late and not worth it now ^^

jkmassel reviewed (approved):
Thanks for addressing the feedback!

I did an end-to-end test with Sentry, which didn't work for me (which lines up with what you'd mentioned!). We'll need to address that (I've left a suggestion), and to test, I'd recommend also adding `-force-crash-logging 1` to both of the app schemes (though switched off by default) to make it possible to test from debug builds.

Then I think this is ready!!

jkmassel reviewed (changes requested):
Accidentally approved instead of requesting changes 🤦

mokagio committed:
Add missing call to start crash logging 🤦‍♂️

Thanks @jkmassel. Oh, the shame...

AliSoftware commented on Simplenote/SimplenoteAppDelegate.m:
```diff
@@ -47,6 +47,8 @@ @interface SimplenoteAppDelegate () <SPBucketDelegate>
 @property (strong, nonatomic) SPUStandardUpdaterController      *updaterController;
 #endif
 
+@property (strong, nonatomic) CrashLogging                      *crashLogging;
```
![spaaaaace (Portal 2 meme)](https://64.media.tumblr.com/tumblr_m55pfm54011r13ra2.gif)

mokagio replied:
😆

AliSoftware reviewed (approved):
Was not able to test on an actual M1, so could only test on Intel and by reviewing code in GitHub UI, but LGTM 👍

mokagio commented:
Adding the `start` call solved the issue.

Here's the demo crashes I sent

<img width="1558" alt="Screen Shot 2021-01-29 at 6 04 52 am" src="https://user-images.githubusercontent.com/1218433/106186524-8371f780-61f8-11eb-97ae-ab6de4536f86.png">

_The bottom one, with the blue dot next to it, is not mine, it's a test I guess @jkmassel sent. It actually confused me for a while because I though it was how Sentry reported its force-crash; I kept reloading its detail page and wondering why mine weren't appearing._

Here's the detail from one
<img width="1559" alt="Screen Shot 2021-01-29 at 6 05 28 am" src="https://user-images.githubusercontent.com/1218433/106186566-91c01380-61f8-11eb-940c-3038382fd1c5.png">

Notice how it was triggered by `aboutAction:` and how the release is `42` (bottom right)

mokagio commented:
@jkmassel asking a new review from you, but reverting this to Draft to make sure I merge it only _after_ #803 is sorted out.

mokagio committed:
Extract build phase for Sparkle code signing in dedicated script

This will make it easier to update it

mokagio committed:
Update Sparkle to version 2.x commit 4767ef6 with Apple Silicon

In #803, we discussed using CocoaPods to fetch Sparkle. That's not
possible because we need support for Sandboxing and Notarization and the
version distributed via CocoaPods doesn't include it.

To support Sandboxing and Notarization, we need to use the WIP version
2.x branch. Unfortunately, the only way to use that version of Sparkle
is to fetch it and build it manually (see ./Scripts/update-sparkle.sh).

The commit I checked out was 4767ef60ece7de7404f402c23177f0810bf72418

I considered using a submodule and scripting the build process, but
decided against it because it would add complexity to the Git setup for
a relatively uncommon operation – we don't need to update Sparkle often.

For more background on Sparkle Sandboxing and Notarization, see:

- https://sparkle-project.org/documentation/sandboxing/
- https://github.com/Automattic/simplenote-macos/pull/383
- https://github.com/Automattic/simplenote-macos/pull/481

mokagio committed:
Add script to automate updating Sparkle

mokagio committed:
Add context regarding the Sparkle setup

mokagio committed:
Only attempt to delete the XPC targets folder if there is one

Thanks @jkmassel.

See https://github.com/Automattic/simplenote-macos/pull/824#discussion_r566577414

jleandroperez commented:
Awesome work gentleman!!! 🎉

mokagio committed:
Remove Sparkle settings from App Store target Info.plist

mokagio committed:
Use longer no-output timeout in CircleCI

See inline comments

mokagio committed:
Manually update Sparkle to support Apple Silicon (#824)
